pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/CastStep.java

```diff
@@ -21,25 +21,101 @@
-
   private final SchemaType keySchemaType;
   private final SchemaType valueSchemaType;
 
   @Override
   public void process(TransformContext transformContext) {
-    if (transformContext.getKeySchema() != null) {
-      Object outputKeyObject = transformContext.getKeyObject();
-      Schema<?> outputSchema = transformContext.getKeySchema();
-      if (keySchemaType == SchemaType.STRING) {
-        outputSchema = Schema.STRING;
-        outputKeyObject = outputKeyObject.toString();
-      }
-      transformContext.setKeySchema(outputSchema);
-      transformContext.setKeyObject(outputKeyObject);
+    if (transformContext.getKeySchema() != null
+        && keySchemaType != null
+        && transformContext.getKeySchema().getSchemaInfo().getType() != keySchemaType) {
+      Object value = convertValue(transformContext.getKeyObject(), keySchemaType);
+      transformContext.setKeySchema(toSchema(keySchemaType));
+      transformContext.setKeyObject(value);
+    }
+    if (valueSchemaType != null
+        && transformContext.getValueSchema().getSchemaInfo().getType() != valueSchemaType) {
+      Object value = convertValue(transformContext.getValueObject(), valueSchemaType);
+      transformContext.setValueSchema(toSchema(valueSchemaType));
+      transformContext.setValueObject(value);
+    }
+  }
+
+  private Object convertValue(Object originalValue, SchemaType schemaType) {
+    switch (schemaType) {
+      case BYTES:
+        return ConverterUtil.toBytes(originalValue);
+      case STRING:
+        return ConverterUtil.toString(originalValue);
+      case BOOLEAN:
+        return ConverterUtil.toBoolean(originalValue);
+      case INT8:
+        return ConverterUtil.toByte(originalValue);
+      case INT16:
+        return ConverterUtil.toShort(originalValue);
+      case INT32:
+        return ConverterUtil.toInteger(originalValue);
+      case INT64:
+        return ConverterUtil.toLong(originalValue);
+      case FLOAT:
+        return ConverterUtil.toFloat(originalValue);
+      case DOUBLE:
+        return ConverterUtil.toDouble(originalValue);
+      case DATE:
+        return ConverterUtil.toDate(originalValue);
+      case TIME:
+        return ConverterUtil.toTime(originalValue);
+      case TIMESTAMP:
+        return ConverterUtil.toTimestamp(originalValue);
+      case INSTANT:
+        return ConverterUtil.toInstant(originalValue);
+      case LOCAL_DATE:
+        return ConverterUtil.toLocalDate(originalValue);
+      case LOCAL_TIME:
+        return ConverterUtil.toLocalTime(originalValue);
+      case LOCAL_DATE_TIME:
+        return ConverterUtil.toLocalDateTime(originalValue);
+      default:
+        throw new IllegalStateException("Unexpected value: " + schemaType);
     }
-    if (valueSchemaType == SchemaType.STRING) {
-      transformContext.setValueSchema(Schema.STRING);
-      transformContext.setValueObject(transformContext.getValueObject().toString());
+  }
+
+  private Schema<?> toSchema(SchemaType schemaType) {
```

Comment:
I was thinking if those are worth caching but the seem defined as fields in the interface (which translates to static final). Just wondering what you think?

Reply:
Yes, primitive Schemas are static so they don't need caching.